### PR TITLE
[11.0] mail: ignore alien domain alias

### PR DESCRIPTION
**Pull request by yajo**
Well, this is a long story, let me see if I can summarize it:

1. **The problem:** Odoo has a long standing problem when routing incoming mail: it ignores the domain part. Imagine you have 2 inboxes: `catchall@example.com` and `info@anotherdomain.com`; your main domain is `example.com`; you configured `info` as an alias for a sales team; an email enters into `info@anotherdomain.com`. Result :arrow_right: It is considered an email for the sales team, even when the email was sent to `info@anotherdomain.com` instead of `info@example.com`.
1. **The solution**: It was included upstream in https://github.com/odoo/odoo/commit/e40f3ccad627af5f74be4fe9f73c201447cb6eb1, including tests and a pretty good description of the problem. Now only emails from the `mail.catchall.domain` were considered for aliases.
1. **The new problem:** There was some people using that bug as a feature: https://github.com/odoo/odoo/pull/38459#issuecomment-555598458
1. **The new solution**: https://github.com/odoo/odoo/pull/41042 reverted partially the original solution (only the part that affected the new problem, because it solved some more things), so we went back to the start.
1. **The new new problem:** See point 1.

Odoo claims to be solving this for master, but that isn't still proven, and we still need a solution; the problem is quite nasty, but we don't want to damage people relying on this bug.

It turns out that in https://github.com/odoo/odoo/pull/41042#issuecomment-559482321 @tde-banana-odoo got a pretty good solution: enable the new (fixed) behavior only in databases where a new ICP `mail.catchall.domain.strict` was set to `True`. That solution can still be seen in the PR history; namely it's commit https://github.com/odoo/odoo/commit/8d5dd14e125f5167ea110c6426a0b18c98e57dba. That solution was discarded in favor of fully reverting the original patch.

**So, what am I doing here?**

Basically:

1. Revert the reversion (https://github.com/odoo/odoo/pull/41042).
1. Apply the 1st approach from https://github.com/odoo/odoo/pull/41042: if you want this fixed in your DB, add the ICP.

@OCA will have to maintain this feature until @odoo fixes it in v14 (if they fulfill their promise).

@Tecnativa TT22715

### CODE REVIEW

**Comment by tde-banana-odoo** [comment]
Hello,

What I could try to make land to facilitate your job is a PR lik
```
def _message_route_validate_recipient(self, email_to):
    """" Allow custom validation by inheritance """
  return bool(email_to)

email_to_localparts = [
    e.split('@', 1)[0].lower()
    for e in (tools.email_split(email_to) or [''])
    if self._message_route_validate_recipient(e)
]
```

That way we simply add a hook in code that is not costly and allow custom behavior on incoming email validation.

**Comment by yajo** [comment]
That'd help, yes.

BTW, Do you have news about the master fetchmail refactoring?

**Comment by yajo** [comment]
#914 was merged for v12, should we merge here too?

**Comment by pedrobaeza** [comment]
Let's go for it

/ocabot merge
